lib/api/src/js/native.rs

```diff
@@ -64,11 +66,34 @@ macro_rules! impl_native_traits {
             pub fn call(&self, mut store: &mut impl AsStoreMut, $( $x: $x, )* ) -> Result<Rets, RuntimeError> where
             $( $x: FromToNativeWasmType + crate::js::NativeWasmTypeInto, )*
             {
-                let params_list: Vec<JsValue> = vec![ $( JsValue::from_f64($x.into_raw(&mut store))),* ];
-                let results = self.handle.get(store.as_store_ref().objects()).function.apply(
-                    &JsValue::UNDEFINED,
-                    &Array::from_iter(params_list.iter())
-                )?;
+                #[allow(unused_unsafe)]
+                let params_list: Vec<RawValue> = unsafe {
+                    vec![ $( RawValue { f64: $x.into_raw(store) } ),* ]
+                };
+                let params_list: Vec<JsValue> = params_list
+                    .into_iter()
+                    .map(|a| a.as_jsvalue(&store.as_store_ref()))
+                    .collect();
+                let results = {
+                    let mut r;
+                    loop {
+                        r = self.handle.get(store.as_store_ref().objects()).function.apply(
+                            &JsValue::UNDEFINED,
+                            &Array::from_iter(params_list.iter())
+                        );
+                        let store_mut = store.as_store_mut();
+                        if let Some(callback) = store_mut.inner.on_called.take() {
+                            match callback(store_mut) {
+                                Ok(wasmer_types::OnCalledAction::InvokeAgain) => { continue; }
+                                Ok(wasmer_types::OnCalledAction::Finish) => { break; }
+                                Ok(wasmer_types::OnCalledAction::Trap(trap)) => { return Err(RuntimeError::user(trap)) },
+                                Err(trap) => { return Err(RuntimeError::user(trap)) },
+                            }
+                        }
+                        break;
+                    }
```

Comment:
I need extra context on why do we need to do this. Is not obvious and I'm not sure we really want this?

@john-sharratt @theduke

Reply:
Decided to keep it.
Code now contains a TODO notice.